### ISSUE TITLE
to_parquet on datetime.date objects works on 2022.5.2, fails on 2022.6.0

### ISSUE THREAD

**Issue by martinedgefocus** [dask/dask#9211]
```
import dask.dataframe as dd
from datetime import date
ddf = dd.from_pandas(pd.DataFrame({'a':[date.today(), date(2022,6,13)]}), npartitions=1)
ddf.to_parquet("/tmp/p")
```

This works OK on 2022.5.2 but regression in 2022.6.0 stops it working.
Really don't want to have to explicitly specify schemas everywhere, that seems unnecessary.

```
ValueError: Failed to convert partition to expected pyarrow schema:
    `ArrowTypeError("Expected bytes, got a 'datetime.date' object", 'Conversion failed for column a with type object')`

Expected partition schema:
    a: string
    __null_dask_index__: int64

Received partition schema:
    a: date32[day]
    __null_dask_index__: int64
```

**Comment by martinedgefocus** [comment]
I think it's this change, but have not proved it:

commit 73c985c2bed5d61be6d05d67f7f381435fe8d1e2
Author: Jim Crist-Harif <jcristharif@gmail.com>
Date:   Fri May 27 09:42:01 2022 -0500

passing `schema=None` prevents the issue

**Comment by rjzamora** [comment]
>passing schema=None prevents the issue

Right - Prior to #9131, the default behavior was for a separate schema to be extracted for each partition at write time (same as `schema=None`). For your single-partition reproducer, this would always work. However, empty/nullable partitions can lead to problems when the schemas are extracted independently. The new default is to use `_meta_nonempty` to "infer" the schema for all partitions up front, and to require the user to specify the `schema` argument explicitly if/when the data in `_meta_nonempty` does not properly reflect the "real" data.

I can understand your frustration here. It seems to me like the only "complete" solution may need to be at the `_meta`/`meta_nonempty` level. However, @ian-r-rose may have other ideas

**Comment by martinedgefocus** [comment]
I think I can understand the motivation ... it does make it substantially harder to use for what should be a very simple operation though. Forcing everyone to explicitly specify schemas everywhere seems undesirable from a usability POV.

I presume the problem case is if someone mixed say datetime.date and str within a column and different partitions could see all date, or all str, and get a conflict? In this case the whole column is just dates, seems a little surprising we can't infer from at least one row or something, rather than always assuming every object is a string?

**Comment by ian-r-rose** [comment]
> I presume the problem case is if someone mixed say datetime.date and str within a column and different partitions could see all date, or all str, and get a conflict?

This is correct -- in general it's really hard to infer what the correct output type is for an `object` dtype. Since most of the time `object`s are string dtypes (at least for the time being), the choice was made to *assume* that `object` meant string dtypes rather than crawl all the partitions and ensure it. This is a more scalable solution, at the cost of some users needing to provide schema information for other `object` types.

But! I think the right solution in your case might be to ensure that your column is a pandas timestamp dtype, rather than messing around with schemas or `meta`:

```python
In [4]: import dask.dataframe as dd
   ...: import pandas as pd
   ...: from datetime import date
   ...: ddf = dd.from_pandas(pd.DataFrame({'a':pd.to_datetime([date.today(), date(2022,6,13)])}), npartitions=1)
   ...: ddf.to_parquet("/tmp/p")
```

would that solve your problem?

**Comment by martinedgefocus** [comment]
Can't do that unfortunately, it's very explicitly a date, not a datetime, else you run into nightmares with timezones, etc.

I totally understand the desire to avoid crawling all partitions, but a sample size of 1 row would seem cheap, and an infinitely better guess than "assume every object is always a string"? If people put mixed types in there it'd still fail, but that seems like a much less common case than any non-string object

**Comment by ian-r-rose** [comment]
Understood, thanks.

To your earlier question: the downside of `schema=None` is that there is no checking that the schemas are at all consistent. So you can run into issues with things like nullable dtypes, as @rjzamora notes above. Even sampling a single row would not necessarily solve that problem. If you are certain that your dataframes wouldn't run into that, then you can feel free to pass `schema=None`.

An alternative would be to provide an explicit schema for your problematic column (which I would probably prefer):

```python
In [13]: import dask.dataframe as dd
    ...: import pandas as pd
    ...: import pyarrow as pa
    ...: from datetime import date
    ...: ddf = dd.from_pandas(pd.DataFrame({'a':[date.today(), date(2022,6,13)]}), npartitions=1)
    ...: ddf.to_parquet("/tmp/p", schema={"a": pa.date32()})
```

**Comment by martinedgefocus** [comment]
Agree that sampling a single row wouldn't necessarily solve the problem, but it does seem like a much better default to me that'd work for most cases without user intervention ?

**Comment by ian-r-rose** [comment]
It's not necessarily as simple as sampling a single row, since that row might be downstream of a lot of computations. If your `to_parquet` call is at the end of a long set of transformations, in order to sample a row, we actually have to do those computations eagerly. Even trying to delay schema inference would force a synchronization point deep in the task graph which causes its own  problems. So we viewed the best compromise as being "any non-pandas-native dtype will probably need to specify the schema".
